api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/MetricsServiceLifecycle.java

```diff
@@ -0,0 +1,241 @@
+
+    private void startMetricsService() {
+        if (state != State.STARTING) {
+            return;
+        }
+        LOG.info("Initializing metrics service");
+        connectionAttempts++;
+        try {
+            session = createSession();
+        } catch (Throwable t) {
+            Throwable rootCause = Throwables.getRootCause(t);
+            LOG.warn("Could not connect to Cassandra cluster - assuming its not up yet", rootCause);
+            // cycle between original and more wait time - avoid waiting huge amounts of time
+            long delay = 1 + ((connectionAttempts - 1) % 4);
+            LOG.warn("[{}] Retrying connecting to Cassandra cluster in [{}]s...", connectionAttempts, delay);
+            lifecycleExecutor.schedule(this::startMetricsService, delay, SECONDS);
+            return;
+        }
+        SchemaManager schemaManager = new SchemaManager(session);
+        try {
+            if (Boolean.parseBoolean(resetDb)) {
+                schemaManager.dropKeyspace(keyspace);
+            }
+            // This creates/updates the keyspace + tables if needed
+            schemaManager.createSchema(keyspace);
+            session.execute("USE " + keyspace);
+            LOG.info("Using a key space of '{}'", keyspace);
+            metricsService.startUp(session);
+            LOG.info("Metrics service started");
+            state = State.STARTED;
+        } catch (Throwable t) {
+            LOG.error("An error occurred trying to connect to the Cassandra cluster", t);
+            state = State.FAILED;
+        }
+    }
+
+    private Session createSession() {
+        Cluster.Builder clusterBuilder = new Cluster.Builder();
+        int port;
+        try {
+            port = Integer.parseInt(cqlPort);
+        } catch (NumberFormatException nfe) {
+            String defaultPort = CASSANDRA_CQL_PORT.defaultValue();
+            LOG.warn("Invalid CQL port '{}', not a number. Will use a default of {}", cqlPort, defaultPort);
+            port = Integer.parseInt(defaultPort);
+        }
+        clusterBuilder.withPort(port);
+        Arrays.stream(nodes.split(",")).forEach(clusterBuilder::addContactPoint);
+
+        Cluster cluster = null;
+        Session session = null;
+        try {
+            cluster = clusterBuilder.build();
+            session = cluster.connect("system");
+            return session;
+        } finally {
+            if (session == null && cluster != null) {
+                cluster.close();
+            }
+        }
+    }
+
+    /**
+     * @return a {@link MetricsService} instance to share in application scope
+     */
+    @Produces
+    @ApplicationScoped
+    public MetricsService getMetricsService() {
+        return metricsService;
+    }
+
+    @PreDestroy
+    void destroy() {
+        Future stopFuture = lifecycleExecutor.submit(this::stopMetricsService);
+        try {
+            Futures.get(stopFuture, 1, MINUTES, Exception.class);
+        } catch (Exception ignore) {
```

Comment:
In general, logging the exception is the right thing to do. But 
sometimes it's really not necessary and I believe it's the case here.

Futures#get takes a checked exception as argument and wraps any Timeout, 
Interrupted or Execution exception with an instance of it. And the code 
which is scheduled for execution is very simple: call 
MetricsService#shutdown and close the C* Session.

It's very much like when you close JDBC result set and connection in a 
finally block, the methods throw a checked Exception but at this stage 
you don't really care.